File: frontend/src/pages/Stain/stainFilters.ts
/**
 * Shared stain classification for the Internal Stain pages.
 *
 * Lives outside StainManagement.jsx so the HosXP Key tab can apply the same
 * rules without importing from its own parent.
 */

export interface StainTest {
  name?: string | null;
  category?: string | null;
  is_external?: boolean | null;
  system_code?: string | null;
}

export interface StainLike {
  is_recut?: boolean | null;
  test?: StainTest | null;
}

export const CAT_COLOR: Record<string, string> = {
  IHC: "purple",
  Histochem: "cyan",
  "Special Stain": "cyan",
  ISH: "geekblue",
  Molecular: "magenta",
};

// Special stains carry two different category strings in the same database:
// the seeded tests use "Histochem", while anything created or edited through
// Admin → master data is saved as "Special Stain" (TEST_CATEGORY_OPTIONS in
// constants/lab.constants.ts). They mean the same thing, so every check has to
// accept both — BlockGridView/StainManagementPage.tsx already does. Compared
// case-insensitively because the backend spells it "Special stain" in places
// (crud/slide_storage.py).
export const SPECIAL_STAIN_CATEGORIES = ["Histochem", "Special Stain"];

export const isSpecialStainCategory = (cat?: string | null): boolean =>
  SPECIAL_STAIN_CATEGORIES.some(
    (c) => c.toLowerCase() === (cat ?? "").toLowerCase(),
  );

/** Short label for the category tag — special stains collapse to "SS". */
export const catLabel = (cat?: string | null): string =>
  isSpecialStainCategory(cat) ? "SS" : cat || "—";

/**
 * The two system tests that are never a stain *order* in their own right: the
 * routine H&E every block gets automatically at grossing (crud/surgical_block
 * create_block) and the H&E recut, which is tracked by `is_recut` and priced
 * at 0. `system_code` is the stable handle — the master test can be renamed
 * from Admin → master data — so the name check is only a fallback for tests
 * created before system_code existed.
 */
export const HE_SYSTEM_CODES = ["HE_ROUTINE", "HE_RECUT"];

export const isRoutineHETest = (test?: StainTest | null): boolean =>
  HE_SYSTEM_CODES.includes(test?.system_code ?? "") ||
  !!test?.name?.includes("H&E");

/** A real in-house special-stain order (AFB, GMS, PAS, …): special-stain
 * category, done in this lab, and not the routine H&E. */
export const isSpecialStainOrder = (s: StainLike): boolean =>
  !s.test?.is_external &&
  isSpecialStainCategory(s.test?.category) &&
  !isRoutineHETest(s.test);

/** What the Internal Stain Orders worklist shows: special stains, plus recuts
 * (BlockGridView routes a pathologist's recut request to this page). IHC / ISH
 * / Molecular are ordered and tracked elsewhere, and the routine H&E has its
 * own page (Stain/RoutineHE). */
export const isRelevantStain = (s: StainLike): boolean =>
  !!s.is_recut || isSpecialStainOrder(s);

/** Billable to HosXP: every in-house test except the H&E pair — a recut is a
 * re-section of an existing block, not a separately charged test. Wider than
 * isRelevantStain on purpose: an in-house IHC still has to be keyed. */
export const isKeyableStain = (s: StainLike): boolean =>
  !s.is_recut && !s.test?.is_external && !isRoutineHETest(s.test);

File: backend/app/utils/stain_filters.py
"""Shared SQL predicates for classifying a SurgicalBlockStain.

The frontend mirror of this file is frontend/src/pages/Stain/stainFilters.ts —
the Internal Stain pages paginate against these predicates on the server and
then render the same rule client-side, so the two must agree.

Every predicate assumes AnatomicalPathologyTest is joined to
SurgicalBlockStain.test_id. Use an OUTER join: a stain's test_id is nullable.
"""

from sqlalchemy import and_, func, or_

from app.models.anatomical_pathology_test import AnatomicalPathologyTest
from app.models.surgical_block_stain import SurgicalBlockStain

# The two system tests that are never a stain *order* in their own right: the
# routine H&E every block gets automatically at grossing (crud.surgical_block
# create_block), and the H&E recut, which is tracked by is_recut and priced at
# 0. system_code is the stable handle — a hospital can rename the master test
# from Admin → master data — so the name check is only a fallback for tests
# created before system_code existed.
HE_SYSTEM_CODES = ("HE_ROUTINE", "HE_RECUT")

# Special stains carry two different category strings in the same database: the
# seeded tests use "Histochem", while anything created through Admin → master
# data is saved as "Special Stain" (TEST_CATEGORY_OPTIONS in the frontend's
# constants/lab.constants.ts). Compared lower-cased because "Special stain"
# also appears in this codebase (crud.slide_storage).
SPECIAL_STAIN_CATEGORIES = ("histochem", "special stain")


def is_not_he_base_test():
    """The stain's master test is neither routine H&E nor an H&E recut.

    Written as a positive NULL-safe test rather than NOT(is H&E) because
    system_code is NULL on every test created from Admin → master data, and
    `NOT (NULL IN (...) OR false)` is NULL, which silently drops the row.
    """
    return and_(
        or_(
            AnatomicalPathologyTest.system_code.is_(None),
            AnatomicalPathologyTest.system_code.notin_(HE_SYSTEM_CODES),
        ),
        AnatomicalPathologyTest.name.notilike("%H&E%"),
    )


def is_special_stain_order():
    """A real in-house special-stain order (AFB, GMS, PAS, …)."""
    return and_(
        AnatomicalPathologyTest.is_external.is_(False),
        func.lower(AnatomicalPathologyTest.category).in_(SPECIAL_STAIN_CATEGORIES),
        is_not_he_base_test(),
    )


def is_internal_stain_order():
    """What the Internal Stain Orders worklist tracks: special stains, plus
    recuts (BlockGridView routes a pathologist's recut request to that page).
    IHC / ISH / Molecular are ordered and tracked elsewhere, and the routine
    H&E has its own page (frontend Stain/RoutineHE)."""
    return or_(SurgicalBlockStain.is_recut.is_(True), is_special_stain_order())


def is_in_house_stain():
    """Any stain this lab runs itself, excluding the H&E pair. Wider than
    is_internal_stain_order on purpose — the HosXP Key tab bills in-house IHC
    too."""
    return or_(
        SurgicalBlockStain.is_recut.is_(True),
        and_(AnatomicalPathologyTest.is_external.is_(False), is_not_he_base_test()),
    )


def is_keyable_stain():
    """Billable to HosXP: every in-house test except the H&E pair — a recut is
    a re-section of an existing block, not a separately charged test."""
    return and_(
        # isnot(True), not is_(False): is_recut is a nullable column, and the
        # frontend's `!s.is_recut` reads NULL as "not a recut" too.
        SurgicalBlockStain.is_recut.isnot(True),
        AnatomicalPathologyTest.is_external.is_(False),
        is_not_he_base_test(),
    )
